src/Language/PureScript/Kinds.hs

```diff
@@ -1,109 +1,164 @@
+{-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveTraversable #-}
 
 module Language.PureScript.Kinds where
 
 import Prelude.Compat
 
 import GHC.Generics (Generic)
 import Control.DeepSeq (NFData)
+import Data.Function (fix)
 import Data.Text (Text)
 import qualified Data.Text as T
-import Data.Aeson.BetterErrors (Parse, key, asText, asIntegral, nth, fromAesonParser, toAesonParser, throwCustomError)
+import Data.Aeson.BetterErrors (Parse, key, asText, asIntegral, nth, fromAesonParser, toAesonParser, throwCustomError, (<|>))
 import Data.Aeson ((.=))
 import qualified Data.Aeson as A
 
+import Language.PureScript.AST.SourcePos
 import Language.PureScript.Names
 import qualified Language.PureScript.Constants as C
 
+type SourceKind = Kind SourceAnn
+
 -- | The data type of kinds
-data Kind
+data Kind a
   -- | Unification variable of type Kind
-  = KUnknown Int
+  = KUnknown a Int
   -- | Kinds for labelled, unordered rows without duplicates
-  | Row Kind
+  | Row a (Kind a)
   -- | Function kinds
-  | FunKind Kind Kind
+  | FunKind a (Kind a) (Kind a)
   -- | A named kind
-  | NamedKind (Qualified (ProperName 'KindName))
-  deriving (Show, Eq, Ord, Generic)
+  | NamedKind a (Qualified (ProperName 'KindName))
+  deriving (Show, Generic, Functor, Foldable, Traversable)
+
+instance NFData a => NFData (Kind a)
+
+srcKUnknown :: Int -> SourceKind
+srcKUnknown = KUnknown NullSourceAnn
+
+srcRow :: SourceKind -> SourceKind
+srcRow = Row NullSourceAnn
 
-instance NFData Kind
+srcFunKind :: SourceKind -> SourceKind -> SourceKind
+srcFunKind = FunKind NullSourceAnn
+
+srcNamedKind :: Qualified (ProperName 'KindName) -> SourceKind
+srcNamedKind = NamedKind NullSourceAnn
 
 -- This is equivalent to the derived Aeson ToJSON instance, except that we
 -- write it out manually so that we can define a parser which is
 -- backwards-compatible.
-instance A.ToJSON Kind where
+instance A.ToJSON a => A.ToJSON (Kind a) where
   toJSON kind = case kind of
-    KUnknown i ->
-      obj "KUnknown" i
-    Row k ->
-      obj "Row" k
-    FunKind k1 k2 ->
-      obj "FunKind" [k1, k2]
-    NamedKind n ->
-      obj "NamedKind" n
+    KUnknown a i ->
+      obj "KUnknown" a i
+    Row a k ->
+      obj "Row" a k
+    FunKind a k1 k2 ->
+      obj "FunKind" a [k1, k2]
+    NamedKind a n ->
+      obj "NamedKind" a n
     where
-    obj :: A.ToJSON a => Text -> a -> A.Value
-    obj tag contents =
-      A.object [ "tag" .= tag, "contents" .= contents ]
+    obj :: A.ToJSON b => Text -> a -> b -> A.Value
+    obj tag ann contents =
+      A.object [ "tag" .= tag, "annotation" .= ann, "contents" .= contents ]
 
 -- This is equivalent to the derived Aeson FromJSON instance, except that it
```

Comment:
I think this comment is now outdated for a similar reason (the `annotation` field vs the `contents` field)